playground/index.ts

```diff
@@ -16,98 +17,176 @@ mermaid.initialize({
+  <pre style="font-size:16px; font-weight:600;font-style:italic;background:#eeeef1;white-space:pre-wrap;width:45vw;padding:5px" id="mermaid-syntax-${id}"></pre>
 
-  <button id="diagram-btn-${i}" data="${i}">Render to Excalidraw</button>
-  <div id="diagram-${i}" style="width:50%"></div>`;
+  <button id="diagram-btn-${id}" data="${id}">Render to Excalidraw</button>
+  <div id="diagram-${id}" style="width:50%"></div>`;
 
-  const btn = diagramContainerEl.querySelector(`#diagram-btn-${i}`)!;
+  const btn = diagramContainerEl.querySelector(`#diagram-btn-${id}`)!;
 
   btn.addEventListener("click", async () => {
     const data = await parseMermaid(diagramDefinition);
+
+    // In HMR we use this attribute to trigger a rerender for active test case for Excalidraw
+    const excalidraw = document.getElementById(EXCALIDRAW_WRAPPER_ID);
+    excalidraw?.setAttribute(EXCALIDRAW_ACTIVE_ATTR, id);
+
     renderExcalidraw(JSON.stringify(data));
   });
 
-  const diagramEl = diagramContainerEl.querySelector(`#diagram-${i}`)!;
-  const { svg } = await mermaid.render(`diagram-${i}`, diagramDefinition);
+  const diagramEl = diagramContainerEl.querySelector(`#diagram-${id}`)!;
+  const { svg } = await mermaid.render(`diagram-${id}`, diagramDefinition);
 
   diagramEl.innerHTML = svg;
   containerEl.append(diagramContainerEl);
 
   // Render mermaid syntax
   const mermaidSyntaxEl = diagramContainerEl.querySelector(
-    `#mermaid-syntax-${i}`
+    `#mermaid-syntax-${id}`
+  )!;
+  mermaidSyntaxEl.innerHTML = diagramDefinition;
+}
+
+async function updateDiagram(
+  name: string,
+  diagramDefinition: string,
+  index: number,
+  baseId: string
+) {
+  const id = generateDiagramId(baseId, index);
+
+  const diagramContainerEl = document.getElementById(
+    `diagram-container-${id}`
   )!;
+  const diagramEl = document.getElementById(`diagram-${id}`)!;
+  const titleEl = document.getElementById(`diagram-title-${id}`)!;
+  const { svg } = await mermaid.render(`diagram-${id}`, diagramDefinition);
+  const mermaidSyntaxEl = document.getElementById(`mermaid-syntax-${id}`)!;
+
+  titleEl.textContent = name;
+  diagramEl.innerHTML = svg;
+  diagramContainerEl.append(diagramEl);
   mermaidSyntaxEl.innerHTML = diagramDefinition;
 }
+
+if (import.meta.hot) {
+  function hmrUpdateTestcase(
+    baseId: string,
+    newTestCases: { name: string; definition: string }[]
+  ) {
+    if (!newTestCases) {
+      return import.meta.hot?.invalidate();
+    }
+
+    for (let i = 0; i < newTestCases.length; i++) {
+      const { name, definition } = newTestCases[i];
+
+      const excalidraw = document.getElementById(EXCALIDRAW_WRAPPER_ID);
+
+      if (
+        excalidraw?.getAttribute(EXCALIDRAW_ACTIVE_ATTR) ===
+        generateDiagramId(baseId, i)
+      ) {
+        parseMermaid(definition).then((data) => {
+          renderExcalidraw(JSON.stringify(data));
+        });
+      }
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏I don't have a strong opinion on this, I think it's best to always re-render excalidraw when the active test has been changed.

Reply:
Excalidraw should be rerendering when elements are getting updated even now, this would trigger the update in the hydrated root which isn't needed